src/librustc_passes/diagnostics.rs

```diff
@@ -248,4 +248,6 @@ register_diagnostics! {
     E0472, // asm! is unsupported on this target
     E0561, // patterns aren't allowed in function pointer types
     E0571, // `break` with a value in a non-`loop`-loop
+    E0583, // `break` with no label in the condition of a `while` loop
+    E0584, // `continue` with no label in the condition of a `while` loop
```

Comment:
For (a), it seems fine to me either way. I'll merge them if you think that'd be simpler.

For (b), I'm happy to write a more extended error explanation (assuming you mean using `register_long_diagnostics!`).